### PR TITLE
[Epic tracker] Drop v1alpha1 support

### PR DESCRIPTION
**Pull request by sairajp-rewind**
#### What this PR does / why we need it:

Removes the `v1alpha1` API version from both API groups (`agents.x-k8s.io` and `extensions.agents.x-k8s.io`), along with the conversion webhook infrastructure that existed solely to serve it. `v1beta1` becomes the only served version for all four CRDs: `Sandbox`, `SandboxClaim`, `SandboxTemplate`, and `SandboxWarmPool`.

**Motivation.** Serving `v1alpha1` costs real CPU even when nobody requests it — the apiserver invokes the conversion webhook to build informer caches regardless of client demand (ref #1234). Beyond that, every `v1beta1` change required a matching lossless conversion, and the self-signed cert machinery, webhook Service, and eight controller flags all existed only to keep `/convert` reachable.

**Targeting `feature/drop-v1alpha1`, not `main`**, per #1265.

#### Epic: Drop deprecated v1alpha1 API support

This PR serves as the **Epic Tracker** for dropping `v1alpha1` support ahead of the v1.0.0 release (Issue #1265). 

To ensure clean, reviewable diffs and independent validation, the changes have been split into **4 focused PRs** targeting `feature/drop-v1alpha1`:

- [x] **PR 1**: `api: drop v1alpha1 spoke packages and regenerate clients` (#1393)
- [x] **PR 2**: `controller, olm: remove conversion webhook infrastructure` (blocked on PR 1) (#1404)
- [x] **PR 3**: `test-migration: remove v1alpha1 to v1beta1 upgrade harness` (#1414)
- [x] **PR 4**: `docs, examples: v1alpha1 removal` (#1427)

---

#### Preserved Discussion & Context
* All previous review threads and architectural Q&A (including OLM lifecycle and upgrade validation) remain recorded here for history.

*(This branch is frozen as discussion substrate; please review the individual child PRs above).*

##### Deliberately unchanged

- Version-agnostic owner-ref checks in `sandboxclaim_controller.go`, `internal/metrics/sandbox_collector.go`, and `sandbox-router/cache/cache.go`, plus their tests — they match on API group rather than version and still guard against legacy owner refs (https://github.com/kubernetes-sigs/agent-sandbox/issues/1265#issuecomment-5064823338).
- `examples/agent-sandbox-rl/tests/test_preflight.py` — `test_crd_wrong_version` uses v1alpha1 as its negative case, which is exactly the realistic stale-cluster scenario the preflight check guards against.
- `site/content/docs/use-cases/anthropic-managed-agents/_index.md` — the GKE Agent Sandbox add-on doesn't serve v1beta1 yet.
- Historical or external-system references in `dev/load-test/test-recipes/README-rapid-burst.md` (a GKE Cluster Autoscaler caching note) and `examples/agent-sandbox-rl/examples/rl_integration.md`.

##### Validation

`make all`, `make test-e2e`, and `./dev/ci/presubmits/test-autogen-up-to-date` all pass. On kind:

- **Fresh install** — CRDs report only `v1beta1`, no webhook Service or cert Secret is created, a v1beta1 Sandbox reconciles to Ready, and all four v1alpha1 kinds are rejected at discovery.
- **Upgrade from v0.5.2** — succeeds when `storedVersions` is pruned to `["v1beta1"]`; correctly rejected by the apiserver when `v1alpha1` remains. All four controllers exercised with the trimmed RBAC, no `Forbidden` errors.
- **GKE** — the latebind storage example ran end-to-end on GKE v1.36.0-gke.3302001 (Filestore CSI + gVisor node pool) against the OSS controller.
- **Helm upgrade path** — verified on kind: install v0.5.2 → prune storedVersions → apply v1beta1-only CRDs → helm upgrade. Helm prunes the webhook Service, Role, and RoleBinding; the runtime-created cert Secret survives (no owner reference, never a chart template) and needs manual deletion. 
-**Upgrade — OLM path (kind + OLM)** — `operator-sdk run bundle` v0.5.2 → `run bundle-upgrade`
  to the new bundle. Verified:
  - the webhook Service carries an `ownerReference` to the CSV
  - upgrading before pruning `storedVersions` is rejected by the apiserver with the error above
  - after pruning, the Service, Role, and RoleBinding are all garbage-collected on CSV replacement
  - the cert Secret is **not** garbage-collected — no owner reference, so OLM has no ownership link

#### Which issue(s) this PR is related to:

Ref #1265

#### Release Note

```release-note
```

## Summary

* **Breaking Changes**
  * `v1alpha1` is removed from the served API surface; `v1beta1` is the only served/supported version.
  * Upgrade requires migrating through `v0.5.x` first for any `v1alpha1`-serialized resources.
  * CRD conversion webhooks and associated migration/conversion tooling are removed.
* **Documentation**
  * Migration guide, API docs, Helm upgrade instructions, and examples updated for `v1beta1`-only behavior, including `status.storedVersions` verification and cleanup steps.
* **Deployment**
  * Webhook service/port/certificates and related RBAC are no longer installed; OLM bundles/CRDs updated accordingly.

### CODE REVIEW

**Comment by GonzaloLuminary** [comment]
For those relying on GKE's add-on, how is this conversion going to be managed? Will GKE automatically migrate the workloads? Is there a timeframe in which GKE will add support for v1beta1?

**Comment by aditya-shantanu** [comment]
> For those relying on GKE's add-on, how is this conversion going to be managed? Will GKE automatically migrate the workloads? Is there a timeframe in which GKE will add support for v1beta1?

in about a month's time. We're trying to do this sooner.

**Comment by sairajp-rewind** [comment]
@janetkuo — `test-e2e-scalability-kwok` is failing here because the script was added to main on 2026-07-27 in #1269, after `feature/drop-v1alpha1` was cut on 07-23. The job config is generated from the scripts present in the checkout, so the feature branch doesn't have it. Not related to this PR — any PR against this base hits it.

Could you merge main into `feature/drop-v1alpha1`? It's 22 commits behind. I'll likely need to resolve conflicts on my branch afterwards.

Keeping this in draft until CI is green post-sync.

**Comment by janetkuo** [comment]
test-e2e-scalability-kwok isn't a required test. You can ignore its failures.

File: docs/api-migration-guide.md [hunk 1]
```diff
@@ -1,338 +1,48 @@
...
-kubectl delete sandboxes,sandboxclaims,sandboxtemplates,sandboxwarmpools -A --all
-kubectl apply -f backup.yaml
-```
-
-Caveats:
-
-- Restoration depends on a functioning conversion webhook. If the webhook itself is broken, fix that first (typically: roll the controller image back to the pre-migration version, then re-apply the backup), or restore in two phases by first re-installing the old chart and then re-applying the backup against the old CRDs.
-- The backup captures `status` subresources too. Strip them before re-apply so the controllers re-derive status from spec rather than racing your stale snapshot: `yq 'del(.items[].status)' backup.yaml | kubectl apply -f -`.
-- Backups don't capture cluster-scoped state like `SandboxWarmPool` controller progress; freshly-applied pools will repopulate themselves from the template.
+The `v1alpha1` API was removed in `vX.Y`.
```
[the remark this answers is not in the document]
Comment (sairajp-rewind) [reply docs/api-migration-guide.md]: Fixed [52d2282](https://github.com/kubernetes-sigs/agent-sandbox/pull/1307/commits/52d22821ea93f6c6b93483b555ea8c5463837fd2)

**Comment by esposem** [comment]
Hi @sairajp-rewind thanks for the ping

Answering the questions written in https://github.com/kubernetes-sigs/agent-sandbox/issues/1265#issuecomment-5112280452

> make -C olm bundle mutated two checked-in files unrelated to this change:

> * olm/config/manager/kustomization.yaml — kustomize edit set image controller=$(IMG) (Makefile:302) rewrote the image from the controller:latest placeholder to a pinned registry.k8s.io/...:v0.5.2. Since the committed bundle CSV already carried the resolved image while config/ kept the placeholder, I read that as the intended convention and reverted. Correct?

I usually don't commit it, it will be overwritten anyways by kustomize so I usually discard this

> * olm/bundle/tests/scorecard/config.yaml — my operator-sdk is newer and emitted storage.spec.mountPath blocks. Reverted as unrelated drift — but it does mean the committed bundle was generated with an older operator-sdk than mine. Worth pinning the version in the Makefile?

I honestly also get this storage.spec.mountPath blocks. I think one of the various bots checking the PR was complaining about them so I decided to discard them to make it happy. 

operator-sdk generates anyways diffs in the bundle that are useless to commit, like for example changes the build date and so on. Therefore feel free to discard all these changes (date, kustomization.yaml and mountPath blocks).

Also please commit the changes in this PR, but split them in two: one commit for the changes in config/ which actually are reviewable, and another for the make bundle output. Also if you are using an AI to do these changes please ensure the make bundle changes are not hand made but via `make bundle`.

**Comment by esposem** [comment]
Answer to the questions:

> 1. Bundle version — 0.6.0 vs 0.5.3? This is a breaking API removal. Should match whatever release version @janetkuo settles on.

Use the latest, then if this PR is still open once the new one comes out, update it and squash in the bundle

> 2. replaces vs semver mode — force a linear upgrade edge, or let users jump from an older bundle straight to this one?

Stick with semver. The catalog is already built with --mode semver (Makefile:356), no replaces or skipRange exists in the current CSV or annotations, and the channel is alpha. Semver mode lets users jump from any older version directly. For alpha-maturity pre-1.0 software, forcing a linear upgrade path through an intermediate version adds complexity with little benefit — users on alpha accept breaking changes.

> 3. Blocked InstallPlans — clusters with v1alpha1 still in status.storedVersions will have their upgrade refuse to proceed. Verified the apiserver rejection on kind:

> status.storedVersions[0]: Invalid value: "v1alpha1": missing from spec.versions;
v1alpha1 was previously a storage version, and must remain in spec.versions until
a storage migration ensures no data remains persisted in v1alpha1 and removes
v1alpha1 from status.storedVersions

The InstallPlan will transition to Failed, not silently stall. The Subscription's .status.conditions will contain the API server rejection message. But the actual error text is buried in condition details — the OperatorHub UI just shows "install failed" without the storedVersions explanation front and center.

  A release-note callout is not enough. Users need explicit pre-upgrade steps:

  1. Migrate all v1alpha1 resources to v1beta1 (kubectl get sandboxes -o yaml, re-apply)
  2. Patch storedVersions on each CRD: `kubectl patch crd sandboxes.agents.x-k8s.io --type=json -p='[{"op":"replace","path":"/status/storedVersions","value":["v1beta1"]}]' --subresource=status`
  3. Then upgrade

  This should go in the migration guide and the OLM README's install section. The OLM upgrade won't auto-retry after the CRD rejection — the user has to clear storedVersions then either re-trigger the InstallPlan or uninstall/reinstall.

> Does OLM surface that usefully, or does the InstallPlan just stall? Release-note callout enough, or something louder?

> 5. hack/strip-csv-webhookdefs (Makefile:305) — strips conversion-webhook definitions from the generated CSV. Dead now, or is it also handling something OLM injects independently?

Dead now

> 6. Does removing a Service and RBAC from a CSV clean up an already-installed operator, or do those objects orphan? On the kubectl path they definitely orphan — I verified four leftover objects and added a cleanup snippet to the migration guide. Want to know whether OperatorHub users need the same.

OLM should handle it, unlike the kubectl path:

  - RBAC (the webhook-certs Secret rules in spec.install.spec.permissions): OLM generates Role/RoleBinding from the CSV's permissions section. On upgrade, it reconciles the Role to match the new CSV — removed rules get dropped. The old CSV's RBAC is garbage-collected when the old CSV reaches  Replaced → deleted.
  - Webhook Service (agent-sandbox-webhook-service_v1_service.yaml): This is a standalone bundle manifest. OLM sets an ownerReference on it pointing to the CSV. When the new bundle doesn't include this file, OLM doesn't adopt it into the new CSV's ownership — it stays owned by the old CSV and gets garbage-collected when the old CSV is cleaned up.

  So OperatorHub users should not need the manual cleanup snippet. That said, OLM's GC of cluster-scoped resources owned by namespace-scoped CSVs has historically had edge cases, and if anyone manually stripped ownerReferences the objects will orphan. Worth verifying on a test cluster with operator-sdk run bundle upgrade, but the expected behavior is clean removal.

File: docs/api-migration-guide.md [hunk 2]
```diff
@@ -1,338 +1,48 @@
...
+  role/agent-sandbox-controller \
+  rolebinding/agent-sandbox-controller \
+  --ignore-not-found
+```
+
+> [!WARNING]
+> The cluster-scoped `ClusterRole` and `ClusterRoleBinding` named `agent-sandbox-controller` are still actively required by the running controller. Do **not** delete cluster-scoped roles; ensure the cleanup snippet above uses namespaced `role`/`rolebinding` within `-n agent-sandbox-system`.
+
+> [!NOTE]
+> **Helm Users**: You do not need to run this cleanup command. Helm natively tracks release manifest ownership and automatically prunes these four removed resources during `helm upgrade`.
```
[the remark this answers is not in the document]
Comment (sairajp-rewind) [reply docs/api-migration-guide.md]: Keeping the blank line: GitHub alert syntax requires separate blockquotes, and > here would merge them so [!NOTE] renders as literal text inside the WARNING box. Happy to merge the two alerts into one if MD028 compliance is preferred.

File: examples/latebind-storage-gke-sandbox/README.md [hunk 3]
```diff
@@ -77,7 +77,7 @@ gcloud beta container clusters update ${CLUSTER_NAME} \
 --enable-agent-sandbox
 ```
 
-**Note**: The GKE add-on currently serves the `v1alpha1` API. The examples in this guide use `v1alpha1` for compatibility with this managed service.
+**Note**: The GKE add-on currently serves `v1alpha1`. If using the GKE add-on, adapt the example manifests from `v1beta1` to `v1alpha1` as needed until the add-on is updated.
```
[the remark this answers is not in the document]
Comment (sairajp-rewind) [reply examples/latebind-storage-gke-sandbox/README.md]: Good catch, the note now spells out that SandboxClaim needs spec.warmPoolRef → spec.sandboxTemplateRef (pointing at the template rather than the pool), not just an apiVersion bump. Didn't ship parallel v1alpha1 manifests (two sets to maintain), but the adaptation steps are now explicit.

File: olm/README.md [hunk 4]
```diff
@@ -19,6 +19,28 @@ This directory packages Agent Sandbox for **OLM** (OperatorHub, OpenShift, and o
...
+       sandboxclaims.extensions.agents.x-k8s.io \
+       sandboxtemplates.extensions.agents.x-k8s.io \
+       sandboxwarmpools.extensions.agents.x-k8s.io; do
+     kubectl patch crd "${crd}" --type=json -p='[{"op":"replace","path":"/status/storedVersions","value":["v1beta1"]}]' --subresource=status
+   done
+   ```
+
+> [!WARNING]
+> **InstallPlan Failure Behavior**: If an OLM upgrade is attempted before `storedVersions` is pruned, the Kubernetes apiserver will reject the CRD update and the OLM `InstallPlan` will transition to **Failed**. In OperatorHub UI, this may appear generically as "install failed" (with the root rejection message visible in the `Subscription` status conditions). OLM will **not** automatically retry after rejection—once you patch `storedVersions`, you must re-trigger/approve the `InstallPlan` or reinstall the subscription.
+
```
[the remark this answers is not in the document]
Comment (sairajp-rewind) [reply olm/README.md:40]: Keeping the blank line: GitHub alert syntax requires separate blockquotes, and > here would merge them so [!NOTE] renders as literal text inside the WARNING box. Happy to merge the two alerts into one if MD028 compliance is preferred.

File: docs/api-migration-guide.md [hunk 5]
```diff
@@ -1,338 +1,58 @@
...
-The other two CRDs (`SandboxTemplate`, `SandboxWarmPool`) are structurally identical between versions but still need a storage rewrite so etcd holds them in v1beta1 form.
+Helm does not upgrade CRDs in `crds/` (see [helm/README.md](../helm/README.md)), so the order matters. Apply the CRDs **before** `helm upgrade`, otherwise the chart removes the webhook Service while the old CRDs still reference it for conversion.
 
-## Two phases
+After completing Migration Steps 1–3 above:
 
-The migration script executes in two distinct phases. **Neither phase can happen in an arbitrary order**, and if you have existing cold-start claims, skipping the bootstrap phase will immediately break them upon upgrading to `v0.5.2`.
+1. `kubectl apply -f helm/crds/` — installs the v1beta1-only CRDs and drops the conversion config.
+2. `helm upgrade` — now safe; removes the webhook Service, Role, and RoleBinding.
+3. Delete the orphaned cert Secret (see Post-Upgrade Cleanup below).
```
[the remark this answers is not in the document]
Comment (sairajp-rewind) [reply docs/api-migration-guide.md:34]: Verified this empirically; installed v0.5.2 via the chart on kind, pruned storedVersions, applied the v1beta1-only CRDs, then helm upgrade. Helm removed the Service, Role, and RoleBinding; only the runtime-created cert Secret survived (no owner reference, never a chart template), which is what prompted the correction in this revision.
Added a verification step plus a pointer to the full cleanup command as a fallback. Keeping the statement that Helm prunes the three manifest-managed resources, since that's what was observed.

File: olm/README.md [hunk 6]
```diff
@@ -19,6 +19,28 @@ This directory packages Agent Sandbox for **OLM** (OperatorHub, OpenShift, and o
...
+> [!NOTE]
+> **Automatic Cleanup**: Unlike standard `kubectl apply` upgrades, OLM automatically garbage-collects legacy webhook `Service` and namespaced `Role`/`RoleBinding` objects when the previous CSV is replaced. Manual deletion of orphaned webhook resources is not required for OLM installations.
```
[the remark this answers is not in the document]
Comment (sairajp-rewind) [reply olm/README.md]: Verified on kind with OLM rather than qualifying: the webhook Service carries an ownerReference to the CSV, and after operator-sdk run bundle-upgrade the Service, Role, and RoleBinding are all garbage-collected. Kept a verification command in the note since clusters vary.
One correction the comment didn't cover: Secret/agent-sandbox-webhook-certs isn't garbage-collected at all, the controller creates it at runtime with no owner reference, so OLM has no ownership link. The note now calls that out with a manual delete command.

**Comment by esposem** [comment]
LGTM, thanks for applying these fixes. two minor nits:
1. tests [presubmit-agent-sandbox-lint-olm](https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/kubernetes-sigs_agent-sandbox/1307/presubmit-agent-sandbox-lint-olm/2082995752728006656) and [presubmit-agent-sandbox-test-olm-unit](https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/kubernetes-sigs_agent-sandbox/1307/presubmit-agent-sandbox-test-olm-unit/2082995753403289600)  are failing simply because you need to rebase, so please do
2. olm/makefile bump should be ideally done in another PR. If this merges soon, then fine, otherwise I can open a new one and bump the makefile already so you don't have to.

**Comment by esposem** [comment]
@sairajp-rewind  my PR to update the makefile has been merged: https://github.com/kubernetes-sigs/agent-sandbox/pull/1330 please rebase

File: docs/api-migration-guide.md [hunk 7]
```diff
@@ -1,338 +1,58 @@
 # v1alpha1 → v1beta1 API migration guide
 
-This document covers the operational side of migrating `Sandbox`, `SandboxClaim`, `SandboxTemplate`, and `SandboxWarmPool` resources from the `v1alpha1` API to the `v1beta1` API.
+The `v1alpha1` API was removed in `v1.0.0`.
```
Comment (aditya-shantanu) [inline docs/api-migration-guide.md:3]: This says v1alpha1 was removed in `v1.0.0`, but `olm/Makefile` and the generated bundle in this same PR are versioned `0.5.3`. Users following this guide to upgrade to `v1.0.0` won't match what OperatorHub publishes as `v0.5.3`; the version reference needs to be consistent across the guide, OLM README, and bundle before this merges.

**Review by aditya-shantanu — CHANGES_REQUESTED** [review]
Reviewed the v1alpha1 removal. The change is structurally sound — conversion webhook infrastructure, CRD v1alpha1 versions, v1alpha1 packages, and related RBAC/Helm templates are all consistently removed. One version inconsistency found; flagging inline.

**Review by janetkuo — COMMENTED** [review]
@sairajp-rewind as discussed, let's break this up into smaller PRs for faster reviews. 

I've also updated the feature branch to pick up the latest change from main.

**Comment by janetkuo** [comment]
/close

given that this was implemented in smaller PRs into the feature branch
